### PR TITLE
json adapters

### PR DESCRIPTION
**Pull request by mjambon**
Still working on it. Feature was implemented and basic test passes.

I'll leave comments next to the interesting changes if you want to take an early peek.

At least I still have to fix merge conflicts and emulate the tag_field feature which I removed.

I probably won't touch this again until next weekend (Apr 28-29). Let's minimize changes in `oj_emit.ml` in the meantime.

### CODE REVIEW

File: atdgen-runtime/src/oj_run.ml [hunk 1]
```diff
@@ -219,6 +219,20 @@ let missing_fields p bit_fields field_names =
...
+  let lb' = Lexing.from_string s' in
+  reader p lb'
+
+let write_with_adapter restore writer ob x =
+  let ob_tmp = Bi_outbuf.create 1024 in
+  writer ob_tmp x;
+  let s_tmp = Bi_outbuf.contents ob_tmp in
+  let ast = Yojson.Safe.from_string s_tmp in
+  let ast' = restore ast in
+  Yojson.Safe.to_outbuf ob ast'
```
Comment (mjambon) [inline atdgen-runtime/src/oj_run.ml]: These two runtime functions apply the user-defined functions `normalize` and `restore` to the json AST. This comes with the cost of an extra read+write.

The big advantage for the user is that `ast` and `ast` have the same, generic type which allows the `normalize` and `restore` functions to be defined independently from the user's atd types.

File: atdgen/src/json.ml [hunk 2]
```diff
@@ -33,7 +37,7 @@ type json_repr =
   | Option
   | Record of json_record
   | String
-  | Sum
+  | Sum of json_adapter option
```
Comment (mjambon) [inline atdgen/src/json.ml]: At this stage, only atd variant types support the `<json adapter=...>` annotation.

It may be useful to support it on records as well; see https://github.com/mjambon/atd/issues/57 for an example.

File: atdgen/src/json.ml [hunk 3]
```diff
@@ -78,13 +124,14 @@ let get_json_fname default an =
...
   Atd.Annot.get_flag ["json"] "keep_nulls" an
 
 let get_json_record an =
   {
     json_keep_nulls = get_json_keep_nulls an;
   }
+
+let tests = [
+  "parse adapter name", test_parse_adapter_name;
+]
```
Comment (mjambon) [inline atdgen/src/json.ml]: to do: run this test suite with the other test.

File: atdgen/test/adapter.ml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  (* Add a `"type"` field. *)
+  let restore (x : json) : json =
+    let cons, assoc =
+      match x with
+      | `List [ `String cons; `Assoc l ] -> cons, l
+      | _ -> assert false
+    in
+    `Assoc (("type", `String (String.lowercase_ascii cons)) :: assoc)
+end
```
Comment (mjambon) [inline atdgen/test/adapter.ml]: This is only used for a test at the moment but I want to add this and more patterns to the atdgen runtime that users could use without hassle.

File: atdgen/test/test3j_t.expected.mli [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  patch1: int option option;
+  patch2: int option option;
+  patch3: int option option
+}
+
+type b = { thing: int }
+
+type a = { thing: string; other_thing: bool }
+
+type adapted = [ `A of a | `B of b ]
```
Comment (mjambon) [inline atdgen/test/test3j_t.expected.mli]: The 4 `test3j_[tj].ml*` files were not checked in, presumably because they were git-ignored.

File: atdgen/src/oj_emit.ml [hunk 6]
```diff
@@ -35,6 +35,7 @@ type param = {
 
 }
 
+let unopt = function None -> assert false | Some x -> x
```
Comment (rgrinberg) [inline atdgen/src/oj_emit.ml]: This is `Option.value_exn` in Import btw.

File: atdgen/src/oj_emit.ml [hunk 7]
```diff
@@ -554,89 +439,48 @@ let rec make_writer p (x : Oj_mapping.t) : Indent.t list =
...
+          Line (sprintf "Bi_outbuf.add_char ob %C" cl);
+        ]
       ]
 
 and make_record_writer p a record_kind =
+  let dot =
+    match record_kind with
+        Record -> "."
+      | Object -> "#"
+  in
```
Comment (rgrinberg) [inline atdgen/src/oj_emit.ml]: You can just use `Ocaml.dot` here.

**Comment by rgrinberg** [comment]
Just to confirm, we aren't breaking the github library in any way here? It's a fairly important consumer of this library.

**Comment by mjambon** [comment]
I am breaking the GitHub library at the moment. The Type_and_value_fields
submodule will help with porting, but it requires a functor application for
each new pair of type/value fields. I haven't tested it yet. I'm hesitating
to preserve syntactic support for the tag_field annotation.

File: atdgen/src/oj_emit.ml (hunk 6 above)
Comment (mjambon) [reply atdgen/src/oj_emit.ml]: Fixed, thanks. I might have brought this old code back when undoing the tag_field implementation.

File: atdgen/src/oj_emit.ml (hunk 7 above)
Comment (mjambon) [reply atdgen/src/oj_emit.ml]: Fixed.

File: atdgen/test/test_atdgen_main.ml [hunk 8]
```diff
@@ -955,41 +584,29 @@ let all_tests = [
...
+(* TODO: use Alcotest to run the test suite. *)
 let quality_test () =
   List.iter (fun f ->
     try f ()
     with Failed -> ())
     all_tests;
   match List.rev !errors with
-  | [] -> ()
+  | [] ->
+      printf "Success.\n%!"
```
[the remark this answers is not in the document]
Comment (mjambon) [reply atdgen/test/test_atdgen_main.ml]: Makes sense. Reverted.

File: ocamlflags.sexp [hunk 9]
```diff
@@ -1 +1,10 @@
...
\ No newline at end of file
+(:standard
+
+  ; Compiler errors and warnings
+  -w -40-43+32+33        ; default set of warnings -this +that
+  -safe-string
+  -short-paths
+  -strict-sequence
+  -strict-formats
+  -keep-locs
```
[the remark this answers is not in the document]
Comment (mjambon) [reply ocamlflags.sexp]: I'm not sure anymore how I came up with this. I simplified the flags and indicated my intent in the file.

File: atdgen/test/test3j.atd [hunk 10]
```diff
@@ -14,122 +14,31 @@ type unixtime_list = int <ocaml repr="float"> list
...
-type fallback_constr = [
-| A <json name="a">
-| Other <json untyped> of (string * json option)
-]
+   See normalize/restore functions in `adapter.ml`.
+*)
+type adapted = [
+  | A <json name="a"> of a
+  | B <json name="b"> of b
+] <json adapter.ocaml="Atdgen_runtime.Util.Json.Adapter.Type_field.Default">
```
Comment (mjambon) [inline atdgen/test/test3j.atd]: ^^^ this is how the new feature is used. We can now have dots in annotation field names. I'm worried about the length of this module path. Is this bad? acceptable?

**Comment by mjambon** [comment]
I  made the required [changes to ocaml-github](https://github.com/mjambon/ocaml-github/tree/atd-adapter) so it works when `tag_field` goes away.

I still need to add back support for `<json untyped>` which is also used a lot in `github.atd`. I can't think of a good alternative. It should be straightforward to implement, though; not sure why I removed it the other day.

**Comment by mjambon** [comment]
After a bit of pondering, I think it would be good reimplement only part of the `<json untyped>` feature. In the case of a pure enum&em;the most common case&em; we had this:
```ocaml
type t = [
  | A
  | Bc
  | Other <json untyped> of (string * json option)
]
```
Now instead we'd write this:
```ocaml
type t = [
  | A
  | Bc
  | Other of string
] <json open_enum>
```
This requires the json elements to be simple strings, and those that are neither `"A"` nor `"Bc"` end up under `Other`. An important difference is that the atd tree is no longer dependent on the `json` type (the argument of `Other` being now just a string).

This feature only covers enums. Variants with arguments would need to be dealt with by writing a custom adapter. This is a bit of work, but such situations are expected to be less common.

**Comment by rgrinberg** [comment]
Martin Jambon <notifications@github.com> writes:

> +type adapted = [
> +  | A <json name="a"> of a
> +  | B <json name="b"> of b
> +] <json adapter.ocaml="Atdgen_runtime.Util.Json.Adapter.Type_field.Default">

The Util module isn't particularly descriptive, so if there's an
opportunity for simplification here, I think it would start by splitting
that module. But otherwise, this looks fine.

**Comment by mjambon** [comment]
@rgrinberg 
`<json adapter.ocaml="Atdgen_runtime.Util.Json.Adapter.Type_field.Default">` is now `<json adapter.ocaml="Atdgen_runtime.Json_adapter.Type_field">`.

**Comment by mjambon** [comment]
@rgrinberg I think we can merge this. Since last time you posted here, I added the `<json open_enum>` feature. I also made changes to ocaml-github to make it work again.

**Comment by rgrinberg** [comment]
Awesome. Thanks. Give me a day to give it one more read and I will give
this a merge if you don't mind.

File: Makefile [hunk 11]
```diff
@@ -1,10 +1,10 @@
 JBUILDER ?= jbuilder
 
 all:
-	$(JBUILDER) build --dev
+	$(JBUILDER) build
 
 tests:
-	$(JBUILDER) runtest --dev
+	$(JBUILDER) runtest
```
Comment (rgrinberg) [inline Makefile]: Why did you remove `--dev` here? You don't like some of the warnings?

File: atdgen/src/json.ml [hunk 12]
```diff
@@ -62,9 +86,38 @@ let get_json_float an : json_float =
...
+   --> { ocaml_adapter = Some "Foo.Bar";
+         java_adapter = None; }
+*)
+let get_json_adapter an =
+  let ocaml_adapter =
+    Atd.Annot.get_field (fun s -> Some (Some s))
+      None ["json"] "adapter.ocaml" an
+  in
+  let java_adapter =
+    Atd.Annot.get_field (fun s -> Some (Some s))
```
Comment (rgrinberg) [inline atdgen/src/json.ml]: Can you remind me why we need the double option type here (and other places)?

File: Makefile (hunk 11 above)
Comment (mjambon) [reply Makefile]: Yes. I remember being annoyed by warning 27. There may be others.
> 27 Innocuous unused variable: unused variable that is not bound with let nor as, and doesn’t start with an underscore (_) character.

I want to write patterns like `Node (loc, x, _)` without having to write `_loc` in case I don't use `loc` later. I don't want to write `Node (_, x, _)` either because the day I need to produce an error that uses `loc`, I don't know where to find it. I also don't want in the same pattern-matching to haves cases like `Node (_loc, x, _)` and other cases like `Other (loc, y)`.

File: atdgen/src/json.ml (hunk 12 above)
Comment (mjambon) [reply atdgen/src/json.ml]: I know it's difficult to make sense of it. I'll try to clarify this with commented examples.

File: atdgen/src/oj_emit.ml [hunk 13]
```diff
@@ -124,120 +118,44 @@ let get_assoc_type deref loc x =
...
-  a
+  List.map (
+    fun x ->
+      let ocaml_fname, ocaml_default, json_fname, optional, unwrapped =
+        match x.f_arepr, x.f_brepr with
+        | Ocaml.Repr.Field o, Json.Field j ->
+            let ocaml_default =
+              match x.f_kind with
+                With_default ->
+                  (match o.Ocaml.ocaml_default with
```
[the remark this answers is not in the document]
Comment (mjambon) [reply atdgen/src/oj_emit.ml]: It's very likely because I undid features by restoring old code from before those features. I'll take a look at your cleanup commits and try to re-apply them.

**Comment by rgrinberg** [comment]
I see. The issue is that that this warning is quite useful in other
situations. There's a way to turn it off selectively with annotations, but
I think then we lose portability with 4.02. I guess we can ignore it for
now.

File: atdgen/src/oj_emit.ml (hunk 13 above)
[the remark this answers is not in the document]
Comment (mjambon) [reply atdgen/src/oj_emit.ml]: ok, I hand-restored your improvements to the `get_fields` function.

File: atdgen/src/json.ml (hunk 12 above)
Comment (mjambon) [reply atdgen/src/json.ml]: I created `Atd.Annot.get_opt_field` and added labeled arguments for clarity.
